plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Future.hs

```diff
@@ -73,230 +151,453 @@ their margin. If either party fails to make a margin payment then the contract
+    ftMarginPenalty + ftUnitPrice
+
+{-# INLINABLE initialState #-}
+-- | The initial state of the 'Future' contract
+initialState :: Future -> FutureState
+initialState ft =
+    let im = initialMargin ft in
+    Running (MarginAccounts{ftsShortMargin=im, ftsLongMargin=im})
+
+futureAddress :: Future -> FutureOwners -> Address
+futureAddress ft fo = Ledger.scriptAddress (validatorScript ft fo)
+
+-- | The data needed to initialise the futures contract.
+data FutureSetup =
+    FutureSetup
+        { shortPK :: PubKey
+        -- ^ Initial owner of the short token
+        , longPK :: PubKey
+        -- ^ Initial owner of the long token
+        , contractStart :: Slot
+        -- ^ Start of the futures contract itself. By this time the setup code 
+        --   has to be finished, otherwise the contract is void.
+        } deriving (Haskell.Show)
+
+-- | Initialise the contract by
+--   * Generating the tokens for long and short
+--   * Setting up an escrow contract for the initial margins
+--   * Paying the initial margin for the given role
+initialiseFuture
+    :: ( HasEndpoint "initialise-future" (FutureSetup, Role) s
+       , HasBlockchainActions s
+       , AsFutureError e
+       )
+    => Future
+    -> Contract s e FutureOwners
+initialiseFuture future = do
+    (s, ownRole) <- endpoint @"initialise-future" @(FutureSetup, Role)
+    ftos <- setupTokens
+    let escr = escrowParams future ftos s
+        escrowPayment = Escrow.payRedeemRefund escr (initialMargin future <> tokenFor Long ftos <> tokenFor Short ftos)
+    e <- withContractError (review _EscrowFailed) escrowPayment
+    either (throwing _EscrowRefunded) (\_ -> pure ftos) e
+
+settleFuture
+    :: ( HasEndpoint "settle-future" (OracleValue Value) s
+       , HasBlockchainActions s
+       , AsFutureError e
+       )
+    => Future
+    -> FutureOwners
+    -> Contract s e ()
+settleFuture future@Future{ftDeliveryDate} ftos = do
+    ov <- endpoint @"settle-future"
+    let address = futureAddress future ftos
+    unspentOutputs <- utxoAt address
+    accounts <- currentBalances future ftos
+    let tx = payoutsTx (payouts future accounts (ovValue ov)) ftos
+             Haskell.<> Tx.collectFromScript unspentOutputs (validatorScript future ftos) (Ledger.RedeemerScript $ PlutusTx.toData $ Settle ov)
+                & validityRange .~ Interval.from (succ ftDeliveryDate)
+    void $ writeTx tx
+
+settleEarly
+    :: ( HasEndpoint "settle-early" (OracleValue Value) s
+       , HasBlockchainActions s
+       , AsFutureError e
+       )
+    => Future
+    -> FutureOwners
+    -> Contract s e ()
+settleEarly future@Future{ftDeliveryDate} ftos = do
+    ov <- endpoint @"settle-early"
+    let address = futureAddress future ftos
+        spotPrice = ovValue ov
+        minMargin = requiredMargin future spotPrice
+    unspentOutputs <- utxoAt address
+    MarginAccounts{ftsShortMargin, ftsLongMargin} <- currentBalances future ftos
+    
+    violatingRole <- 
```

Comment:
Is there some duplicate logic with the `check` function here?

Reply:
Possibly this is another iteration of https://github.com/input-output-hk/plutus/issues/1398#issuecomment-527383756, which I still really like. (That comment doesn't say it, but we can also use the constraints on the chain to check that the `PendingTx` satisfies them, which is nice!)

Reply:
> Is there some duplicate logic with the check function here?

Yeah I'll try to merge them

> Possibly this is another iteration of #1398 (comment), which I still really like

It would make a lot of things easier...